File: src/test/java/com/github/natche/gravatarjavaclient/avatar/GravatarAvatarRequestImageSaverTest.java
package com.github.natche.gravatarjavaclient.avatar;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.mock;

import com.github.natche.gravatarjavaclient.exceptions.GravatarJavaClientException;
import com.google.common.collect.ImmutableList;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.awt.image.BufferedImage;
import java.io.File;
import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.List;

/**
 * Tests for the {@link GravatarAvatarRequestImageSaver}.
 */
public class GravatarAvatarRequestImageSaverTest {
    /**
     * Sets the SUPPORTED_IMAGE_FORMATS to an ArrayList containing a blank string which will cause
     * the image write to fail. This is performed for testing coverage purposes.
     */
    @SuppressWarnings("unchecked") /* Reflection; expected */
    @BeforeEach
    public void injectBlankFormat() {
        try {
            Field formatsField = GravatarAvatarRequestImageSaver.class
                    .getDeclaredField("SUPPORTED_IMAGE_FORMATS");
            formatsField.setAccessible(true);
            ImmutableList<String> originalList
                    = (ImmutableList<String>) formatsField.get(GravatarAvatarRequestImageSaver.INSTANCE);
            List<String> mutableList = new ArrayList<>(originalList);
            mutableList.add("");
            formatsField.set(GravatarAvatarRequestImageSaver.INSTANCE, ImmutableList.copyOf(mutableList));
        } catch (Exception e) {
            // Swallow
        }
    }

    /**
     * Tests for the save to method.
     */
    @Test
    void testSaveTo() {
        assertThrows(NullPointerException.class,
                () -> GravatarAvatarRequestImageSaver.INSTANCE.saveTo(null, null, null));

        BufferedImage mockImage = mock(BufferedImage.class);
        assertThrows(NullPointerException.class,
                () -> GravatarAvatarRequestImageSaver.INSTANCE.saveTo(mockImage, null, null));

        File mockFile = mock(File.class);
        assertThrows(NullPointerException.class,
                () -> GravatarAvatarRequestImageSaver.INSTANCE.saveTo(mockImage, mockFile, null));
        assertThrows(IllegalArgumentException.class,
                () -> GravatarAvatarRequestImageSaver.INSTANCE.saveTo(mockImage, new File("."), null));

        assertThrows(GravatarJavaClientException.class,
                () -> GravatarAvatarRequestImageSaver.INSTANCE.saveTo(mockImage, mockFile, ""));
    }
}

File: src/test/kotlin/com/github/natche/gravatarjavaclient/avatar/GravatarAvatarRequestImageSaverTest.kt
package com.github.natche.gravatarjavaclient.avatar

import com.github.natche.gravatarjavaclient.exceptions.GravatarJavaClientException
import com.google.common.collect.ImmutableList
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import java.awt.image.BufferedImage
import java.io.File

/**
 * Tests for the [GravatarAvatarRequestImageSaver].
 */
class GravatarAvatarRequestImageSaverTest {
    /**
     * Sets the SUPPORTED_IMAGE_FORMATS to an ArrayList containing a blank string which will cause
     * the image write to fail. This is performed for testing coverage purposes.
     */
    @BeforeEach
    fun injectBlankFormat() {
        try {
            val formatsField = GravatarAvatarRequestImageSaver::class.java
                .getDeclaredField("SUPPORTED_IMAGE_FORMATS")
            formatsField.isAccessible = true
            @Suppress("UNCHECKED_CAST")
            val originalList = formatsField[GravatarAvatarRequestImageSaver.INSTANCE] as ImmutableList<String>
            val mutableList: MutableList<String> = ArrayList(originalList)
            mutableList.add("")
            formatsField[GravatarAvatarRequestImageSaver.INSTANCE] = ImmutableList.copyOf(mutableList)
        } catch (e: Exception) {
            // Swallow
        }
    }

    /**
     * Tests for the getSavedCount method.
     */
    @Test
    fun testGetSavedCount() {
        assertEquals(0, GravatarAvatarRequestImageSaver.INSTANCE.savedCount)
    }

    /**
     * Tests for the save to method.
     */
    @Test
    fun testSaveTo() {
        assertThrows(
            NullPointerException::class.java
        ) { GravatarAvatarRequestImageSaver.INSTANCE.saveTo(null, null, null) }
        val mockImage = Mockito.mock(BufferedImage::class.java)
        assertThrows(
            NullPointerException::class.java
        ) { GravatarAvatarRequestImageSaver.INSTANCE.saveTo(mockImage, null, null) }
        val mockFile = Mockito.mock(File::class.java)
        Mockito.`when`(mockFile.name).thenReturn("validName.png")

        assertThrows(
            NullPointerException::class.java
        ) { GravatarAvatarRequestImageSaver.INSTANCE.saveTo(mockImage, mockFile, null) }
        assertThrows(
            IllegalArgumentException::class.java
        ) { GravatarAvatarRequestImageSaver.INSTANCE.saveTo(mockImage, File("."), null) }
        assertThrows(
            GravatarJavaClientException::class.java
        ) { GravatarAvatarRequestImageSaver.INSTANCE.saveTo(mockImage, mockFile, "") }
        assertThrows(
            java.lang.IllegalArgumentException::class.java
        ) { GravatarAvatarRequestImageSaver.INSTANCE.saveTo(mockImage, File("invalid<>.png"), "") }
    }
}